### PR TITLE
Fix #216: Crypto 3.2: Test new protocol version

### LINKED ISSUE #216: Crypto 3.2: Test new protocol version
Add tests for version `3.2` of PowerAuth protocol, focusing on ECIES changes.

### PR DESCRIPTION
**Pull request by romanstrobl**

### CODE REVIEW

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/v32/PowerAuthRecoveryTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Autowired
+    public void setPowerAuthTestConfiguration(PowerAuthTestConfiguration config) {
+        this.config = config;
+    }
+
+    @BeforeEach
+    void setUp() throws IOException {
+        // Create temp status file
```
Comment (banterCZ) [inline powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/v32/PowerAuthRecoveryTest.java:60]: Quite obvious.

**Review by banterCZ — APPROVED** [review]
Quite a large PR, not able to review in detail. But let's merge it. Thanks for extracting the shared logic. I have noticed several minor issues, but most probably present in the original code, so we may improve in consecutive PRs.

**Comment by romanstrobl** [comment]
The goal of this pull request was to allow better versioning of tests based on cryptographic protocol version, not to migrate to new Java constructs or clean-up the code. I will implement requested changes, but note that more work may be required, which is out of scope of this pull request.

File: powerauth-backend-tests/src/main/java/com/wultra/security/powerauth/configuration/PowerAuthTestConfiguration.java [hunk 2]
```diff
@@ -305,6 +344,19 @@ public void setActivationIdV3(String activationIdV3) {
...
+        switch (version) {
+            case "3.2":
+                return activationIdV32;
+            case "3.1":
+                return activationIdV31;
+            case "3.0":
+                return activationIdV3;
+            default:
+                return null;
+        }
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-backend-tests/src/main/java/com/wultra/security/powerauth/configuration/PowerAuthTestConfiguration.java:357]: Good idea.

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationOtpShared.java [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+            boolean lastIteration = iteration == 5;
+            CommitActivationRequest commitRequest = new CommitActivationRequest();
+            commitRequest.setActivationId(initResponse.getActivationId());
+            commitRequest.setActivationOtp(invalidOtpValue);
+
+            boolean isActivated;
+            try {
+                CommitActivationResponse commitResponse = powerAuthClient.commitActivation(commitRequest);
+                isActivated = commitResponse.isActivated();
+            } catch (Throwable t) {
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationOtpShared.java:213]: We can remove the catch blocks completely, any exception will fail the test.

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationOtpShared.java [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+        assertTrue(stepLoggerPrepare.getResult().success());
+        assertEquals(200, stepLoggerPrepare.getResponse().statusCode());
+
+        // Verify activation status
+        GetActivationStatusResponse activationStatusResponse = powerAuthClient.getActivationStatus(initResponse.getActivationId());
+        assertEquals(ActivationStatus.PENDING_COMMIT, activationStatusResponse.getActivationStatus());
+
+        // Try commit with no OTP for more than max failed attempts. Use OTP in the last iteration, that should pass.
+        for (int iteration = 1; iteration <= 6; iteration++) {
+            boolean lastIteration = iteration == 6;
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationOtpShared.java:436]: I do not see any benefit, so if you want to change it, you can do it in separate issue.

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationShared.java [hunk 5]
```diff
@@ -180,30 +135,29 @@ void activationPrepareTest() throws Exception {
...
 
         // Verify that application version is unsupported
-        GetApplicationDetailResponse detailResponse = powerAuthClient.getApplicationDetail(config.getApplicationId());
-        for (ApplicationVersion version : detailResponse.getVersions()) {
-            if (version.getApplicationVersionId().equals(config.getApplicationVersion())) {
-                assertFalse(version.isSupported());
+        final GetApplicationDetailResponse detailResponse = powerAuthClient.getApplicationDetail(config.getApplicationId());
+        for (ApplicationVersion appVer: detailResponse.getVersions()) {
+            if (appVer.getApplicationVersionId().equals(config.getApplicationVersion())) {
+                assertFalse(appVer.isSupported());
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationShared.java:153]: Yes, I suppose we can change it.

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/v32/PowerAuthRecoveryTest.java (hunk 1 above)
Comment (romanstrobl) [reply powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/v32/PowerAuthRecoveryTest.java:60]: Less obvious after we added version processing, I will keep the comment, there is nothing wrong about it.

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationCodeShared.java [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        assertNotNull(activationCode);
+        assertNotNull(activationSignature);
+
+        byte[] activationSignatureBytes = Base64.getDecoder().decode(activationSignature);
+
+        // Verify activation signature
+        try {
+            boolean activationSignatureOK = CLIENT_ACTIVATION.verifyActivationCodeSignature(activationCode, activationSignatureBytes, config.getMasterPublicKey());
+            assertTrue(activationSignatureOK);
+        } catch (Throwable t) {
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationCodeShared.java:110]: Removed.

File: powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationCodeShared.java [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        signatureModel.setResourceId("/api/activation/code");
+        signatureModel.setUriString(config.getEnrollmentServiceUrl() + "/api/activation/code");
+
+        File dataFile = File.createTempFile("data_activation_code", ".dat");
+        dataFile.deleteOnExit();
+        BufferedWriter out = Files.newBufferedWriter(dataFile.toPath(), StandardCharsets.UTF_8);
+
+        String additionalActivationOtp = UUID.randomUUID().toString();
+
+        String requestData = "{\n" +
```
[the remark this answers is not in the document]
Comment (romanstrobl) [reply powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/shared/PowerAuthActivationCodeShared.java:68]: There are variables in this String. Again, out of scope of this PR, please raise an issue and fix separately.

**Pull request merged**